### PR TITLE
Implement Fog Ledger Router

### PR DESCRIPTION
**Pull request by awygle**
### Motivation

This PR contains all of the feature work related to the Fog Ledger Router. Each change has already been independently reviewed on the feature/fog-ledger-router branch, and the branch itself has been kept up-to-date with master.

### CODE REVIEW

**Review by samdealy — APPROVED** [review]

**Review by nick-mobilecoin — CHANGES_REQUESTED** [review]
The main blocker is `tempdir` shouldn't come back into the mainline

I'm also curious if there is any discussion the `chain_id` validation for the router, shards, and stores.

File: fog/ledger/server/src/key_image_service.rs [hunk 1]
```diff
@@ -32,7 +36,7 @@ pub struct KeyImageService<L: Ledger + Clone, E: LedgerEnclaveProxy> {
 
 impl<L: Ledger + Clone, E: LedgerEnclaveProxy> KeyImageService<L, E> {
     pub fn new(
-        chain_id: String,
+        client_listen_uri: KeyImageStoreUri,
```
[the remark this answers is not in the document]
Comment (awygle) [reply fog/ledger/server/src/key_image_service.rs:39]: I'm not actually sure what the deal is here. `chain_id` is present in FVR but only for the "legacy" unary API. It seems to be part of the auth stuff we do for unary but not for streaming. FLR has it for the APIs that didn't get router-ized, but doesn't seem to have preserved it for the unary Key Image API path.

File: Cargo.lock [hunk 2]
```diff
@@ -25,9 +19,9 @@ checksum = "f26201604c87b1e01bd3d98f8d5d9a8fcbb815e8cedb41ffccbeb4bf593a35fe"
 
 [[package]]
 name = "aead"
-version = "0.5.1"
+version = "0.5.2"
```
Comment (nick-mobilecoin) [inline Cargo.lock:22]: :wrench: I was hoping most of the cargo lock would get fixed by the `tempdir` removal but there are still 1600+ changes.
Should move this closer to the mainline, something like the following could be done with this branch checked out.
```shell
git checkout origin/master Cargo.lock
cargo check
git add Cargo.lock
git commit -m "Revert unnecessary Cargo.lock changes"
```

**Review by nick-mobilecoin — CHANGES_REQUESTED** [review]
Just the cargo lock causing a lot of churn that doesn't seem necessary

**Review by nick-mobilecoin — APPROVED** [review]

File: fog/ledger/server/src/key_image_service.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply fog/ledger/server/src/key_image_service.rs:39]: created #3320 to track this.

**Pull request merged**